### DISCUSSION TITLE
Have distinct options for build setup instead of having to use `CIBW_BUILD`

### DISCUSSION THREAD

**Discussion by Zuzu-Typ** [pypa/cibuildwheel#925] [Ideas]
In the latest cibuildwheel version (when building on linux), manylinux and musllinux, as well as CPython 3.6-10 and PyPy 3.7 are built by default.
As a quick side note: I couldn't find much information about musllinux (especially since the [repo](https://github.com/pypa/musllinux) linked in the [docs](https://cibuildwheel.readthedocs.io/en/stable/options/#linux-image) doesn't exist or is private).

My library ([PyGLM](https://github.com/Zuzu-Typ/PyGLM)) is not compatible with PyPy, so I have to [use](https://github.com/Zuzu-Typ/PyGLM/blob/master/.github/workflows/deploy.yml#L8) the `CIBW_BUILD` variable to specify CPython versions only.
```
CIBW_BUILD: cp3*
```
In the latest version of cibuildwheel, musllinux is built by default, however, since I want to support all manylinux variants, I build manylinux2010 and manylinux_2_24 wheels (which are backwards-compatible to more manylinux variants) and thus musllinux is built twice.
To avoid this, I would have to specify to only build manylinux using the `CIBW_BUILD` variable (in addition to only building CPython), which is a very janky solution.
```
CIBW_BUILD: cp3*-manylinux*
```
(I haven't tested this)

I think there should be options to specify what to build more easily.
For example, whether to build CPython or PyPy or both, whether to build manylinux or musllinux or both (maybe even multiple variants of them) and which versions of CPython to build (e.g. 3.8 through 3.10).
Example:
```
CIBW_PYTHON_INTERPRETER: CPython
CIBW_CPYTHON_VERSIONS: 3.8 3.9 3.10
CIBW_LINUX_IMAGES: manylinux2010 musllinux_1_1
```

**Comment by henryiii** [comment]
We already get some freedom with ARCHS (auto32 and auto64 are handy), but in general, this selection is done by something in your build matrix - and that is much more complex to set when you have multiple variables, while the one variable is cleaner, and easier to check or use for naming inside the matrix. I would set `[tool.cibuildwheel] skip = "pp*"` to skip pypy permanently, and you can use  requires-python (any of the ways it can be set, ideally in setup.cfg for now) to limit the lower bound of Python to 3.8. Then use BUILD inside your matrix to do the runtime splitting to build manylinux twice and musllinux.

Overrides are also tied to build selectors. Having a "prettier" way to select build selectors with multiple variables would just add a lot of confusion, and very few users would use them. One simple way is better than multiple complex ways. The only thing I'd be tempted to change here is to have a nicer way to set up the "parallel" builds - building multiple manylinuxs is pretty common and I do have an idea for this, but again it's completely tied to the build selectors.

**Reply by Zuzu-Typ** [reply]
> this selection is done by something in your build matrix

Oh. Right. Somehow I didn't think of that 😅
Yeah, when using a placeholder I do see why it's cleaner.



> I would set `[tool.cibuildwheel] skip = "pp*"` to skip pypy permanently

Thanks for the suggestion. 
Are any other interpreters planned?

**Comment by henryiii** [comment]
Also, your method is a lot less powerful. There's no way to specify different Python versions for different Python interpreters, etc - while that's easy with build selectors. This is really important with PyPy, which doesn't support manylinux1 (or soon manylinux2010), and only comes in one (soon two) Python versions.

**Reply by Zuzu-Typ** [reply]
> while that's easy with build selectors

I suppose you mean by using a build matrix, correct?
Or is it possible to specify multiple build selectors?

**Comment by joerick** [comment]
> To avoid this, I would have to specify to only build manylinux using the `CIBW_BUILD` variable (in addition to only building CPython), which is a very janky solution.
> 
> ```
> CIBW_BUILD: cp3*-manylinux*
> ```

There's no accounting for taste, I suppose, but this doesn't seem to janky to me! :P Maybe I'm just used to it.

As @henryiii mentioned, it might be easier to use `CIBW_SKIP=pp*` for the pypy concern, then zero-in on other options using the CIBW_BUILD option.

Perhaps this would be improved by our idea of an interactive selector table, where values of build/skip/requires-python can be tested in the docs?

**Reply by Zuzu-Typ** [reply]
> but this doesn't seem to janky to me!

Yeah, when you don't have to type that in manually and instead have the build matrix do it for you, I agree :D 



> Perhaps this would be improved by our idea of an interactive selector table, where values of build/skip/requires-python can be tested in the docs?

Would this be like a small dialog, saying
_if you set CIBW_BUILD as [...], CIBW_SKIP as [...] and CIBW_ARCHS as [...] (etc.), cibuildwheel will build the following wheels for you: [...]_ 
(or something similar)?

**Reply by joerick** [reply]
Yes, a bit like that. I think @henryiii has a prototype in a codepen somewhere.
